enhancements/hybrid-helm-operator.md

```diff
@@ -0,0 +1,223 @@
+
+## Open Questions
+
+1. How do we handle migrations for existing Helm-based Operators? For the first iteration
+   requiring users to manually migrate charts and watches.yaml changes should be enough,
+   but longer term we may want to support a command to automatically migrate this over,
+   perhaps as an initialization flag to `operator-sdk init`, similar to how the `--helm-chart`
+   argument works for the helm plugin.
+1. Where should the helm library code live? Likely it should either go in `operator-frameowkr/operator-lib`,
+   but it may make sense for it to have its own repository.
+1. How would interaction between Helm and Go apis work? Is it possible to generate a `types.go` for the Helm
+   apis that would allow the Go code to interact with them easily?
+
+## Summary
+
+Allow helm-operator users to extend their operators using Golang. We would accomplish
+this by exposing all the functionality that the helm-operator has via a library
+and scaffolding out code that is equivalent to the default behavior of the helm-operator.
+The user could then modify that code as they see fit, effectively turning their helm-based
+operator into a Golang one, but leaving a few utilities specific to helm-operators available
+for convenience.
+
+
+## Motivation
+
+Helm is quite limited compared to Ansible or Golang, the other two technologies that
+the Operator SDK currently supports. Once a helm-operator author reaches the limits
+of what they can accomplish with Helm and begins to need additional functionality, the
+only choice they have is to throw away what they have and duplicate it with another
+technology. Even then, maintaining a compatible API is non-trivial as the helm-operator
+provides functionality that the user would need to reimplement. Rather than starting over,
+a hybrid operator approach would allow them to implement advanced functionality in
+Golang while leaving the existing helm functionality in place.
+
+### Goals
+
+- Provide a library containing some of the helper logic for interacting with Helm and
+  controller-runtime
+- Scaffold out whatever code does not fit easily with the library
+- If the user changes nothing in the code, their operator should behave exactly as it
+  did when it was a plain helm-operator.
+- The user should be able to scaffold APIs in either Golang or Helm
+- In the end, we are trying to keep the ease of use of Helm-based operators, while adding
+  the extensibility of Golang-based operators.
+
+### Non-Goals
+
+- We will not continue to manage the hybrid operator in any way, once it is converted
+  it is essentially a Go-based operator and so will not make use of a base image with
+  dependencies, automatic upgrades of helm versions, or any other conveniences provided
+  by the base image. The user will need to manage and upgrade their code by hand.
+
+## Proposal
+
+There are 3 stages to this proposal:
+
+1. Joe Lanford has made significant progress on this already, in the
```

Comment:
Another option would be to have a separate repo with _just_ the helm-operator library + plugin code, all of which could be imported from other projects.

SDK would import the plugin and code scaffolded by the plugin would import the library.

Something like this:
```
 _____                    _________
|     |     scaffolds    |         |
| SDK |----------------->| project |
|_____|                  |_________|
  |                               |
 /                                 \
 | imports                         | imports
 |  ____________________________   |
 | |   ________     _________   |  |
 | |  |        |   |         |  |  |
 \-+->| plugin |   | library |<-+--/
   |  |________|   |_________|  |
   |                            |
   |              helm-operator |
   |____________________________|
   
```

The big reason to separate things out this way is that we could keep overall versioning of each hybrid operator type different, which would enable them to evolve at different paces.

Reply:
I think in the long run a separate repo definitely makes sense, but that's kind of a whole other can of worms as how to handle a plugin repo split is still an ongoing discussion. That would be a good reason to break the helm library bits into their own repo though, as it would serve as the eventual home of all the code as the sdk/plugin architecture matures. I think I'd want to avoid tying development work on the hybrid operator to the completion of that discussion though

Reply:
After further thought/discussion I think you're right about moving to the new repo initially, updated the proposal to reflect that.